components/brave_federated/config_utils.cc

```diff
@@ -0,0 +1,133 @@
+
+namespace brave_federated {
+
+LearningServiceConfig::LearningServiceConfig() {
+  reconnect_policy_ = {/*.num_errors_to_ignore = */ 0,
+                       /*.initial_delay_ms = */ 10 * 1000,
+                       /*.multiply_factor =*/2.0,
+                       /*.jitter_factor =*/0.0,
+                       /*.maximum_backoff_ms =*/60 * 10 * 1000,
+                       /*.always_use_initial_delay =*/true};
+  request_task_policy_ = {
+      /*.num_errors_to_ignore = */ 0,
+      /*.initial_delay_ms = */
+      features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      /*.multiply_factor =*/2.0,
+      /*.jitter_factor =*/0.0,
+      /*.maximum_backoff_ms =*/16 *
+          features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      /*.always_use_initial_delay =*/true};
+  post_results_policy_ = {
+      /*.num_errors_to_ignore = */ 0,
+      /*.initial_delay_ms = */
+      features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      /*.multiply_factor =*/2.0,
+      /*.jitter_factor =*/0.0,
+      /*.maximum_backoff_ms =*/16 *
+          features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      /*.always_use_initial_delay =*/true};
+  model_spec_ = {/* .num_params = */ 32,
+                 /* .batch_size = */ 32,
+                 /* .learning_rate = */ 0.01,
+                 /* .num_iterations = */ 500,
+                 /* .threshold = */ 0.5};
+}
+
+LearningServiceConfig::LearningServiceConfig(const base::FilePath& path)
+    : LearningServiceConfig::LearningServiceConfig() {
+  std::string data;
+  const bool success = base::ReadFileToString(path, &data);
+  VLOG(2) << "Data: " << data;
+  if (!success || data.empty()) {
+    VLOG(1) << "Error in reading JSON configuration from " << path;
+    return;  // return default ctor's initialisation
+  }
+
+  InitServiceConfigFromJSONString(data);
+}
+
+LearningServiceConfig::LearningServiceConfig(const std::string data)
+    : LearningServiceConfig::LearningServiceConfig() {
+  InitServiceConfigFromJSONString(data);
+}
+
+void LearningServiceConfig::InitServiceConfigFromJSONString(
+    const std::string data) {
+  const absl::optional<base::Value> root =
+      base::JSONReader::Read(data, base::JSON_PARSE_CHROMIUM_EXTENSIONS |
+                                       base::JSONParserOptions::JSON_PARSE_RFC);
+
+  if (!root || !root->is_dict()) {
+    VLOG(1) << "Error in configuration file: root is not a dict.";
+    return;  // return default ctor's initialisation
+  }
+
+  const base::Value::Dict& dict = root->GetDict();
+
+  base::JSONValueConverter<CustomBackoffEntryPolicy> policy_converter;
+  CustomBackoffEntryPolicy custom_reconnect_policy(reconnect_policy_);
+  CustomBackoffEntryPolicy custom_request_task_policy(request_task_policy_);
+  CustomBackoffEntryPolicy custom_post_results_policy(post_results_policy_);
+
+  bool result = false;
+  result = policy_converter.Convert(*(dict.Find("reconnect_policy")),
```

Comment:
find can return nullptr. Also "reconnect_policy" and other string literals should be constexpr `kReconnectPolicy', etc...

Reply:
Fixed strings. Nullptr seems not not be causing an issue.

Reply:
Theoretically, yes, because of `nullptr` dereference being undefined behaviour. In essence, this was not happening. In any case, I added the explicit checks.

Reply:
Great as always good practice to fail gracefully in such situations. Thanks